src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+end
+
+function copy_model_to_model(state::FirstPassState, from_addr, to_addr)
+    trace_choices = get_choices(state.trace)
+    push!(state.t_copy_reads, from_addr)
+    if has_value(trace_choices, from_addr)
+        state.constraints[to_addr] = state.trace[from_addr]
+    else
+        set_submap!(state.constraints, to_addr, get_submap(trace_choices, from_addr))
+    end
+    nothing
+end
+
+function copy_model_to_model(state::JacobianPassState, from_addr, to_addr)
+    nothing
+end
+
+# copy_model_to_proposal
+
+macro copy_model_to_proposal(from_addr, to_addr)
+    quote copy_model_to_proposal($(esc(inv_state)), $(esc(from_addr)), $(esc(to_addr))) end
+end
+
+function copy_model_to_proposal(state::FirstPassState, from_addr, to_addr)
+    trace_choices = get_choices(state.trace)
+    push!(state.t_copy_reads, from_addr)
+    if has_value(trace_choices, from_addr)
+        state.u_back[to_addr] = state.trace[from_addr]
+    else
+        set_submap!(state.u_back, to_addr, get_submap(trace_choices, from_addr))
+    end
+    nothing
+end
+
+function copy_model_to_proposal(state::JacobianPassState, from_addr, to_addr)
+    nothing
+end
+
+# copy_proposal_to_proposal
+
+macro copy_proposal_to_proposal(from_addr, to_addr)
+    quote copy_proposal_to_proposal($(esc(inv_state)), $(esc(from_addr)), $(esc(to_addr))) end
+end
+
+function copy_proposal_to_proposal(state::FirstPassState, from_addr, to_addr)
+    push!(state.u_copy_reads, from_addr)
+    if has_value(state.u, from_addr)
+        state.u_back[to_addr] = state.u[from_addr]
+    else
+        set_submap!(state.u_back, to_addr, get_submap(state.u, from_addr))
+    end
+    nothing
+end
+
+function copy_proposal_to_proposal(state::JacobianPassState, from_addr, to_addr)
+    nothing
+end
+
+# copy_proposal_to_model
+
+macro copy_proposal_to_model(from_addr, to_addr)
+    quote copy_proposal_to_model($(esc(inv_state)), $(esc(from_addr)), $(esc(to_addr))) end
+end
+
+function copy_proposal_to_model(state::FirstPassState, from_addr, to_addr)
+    push!(state.u_copy_reads, from_addr)
+    if has_value(state.u, from_addr)
+        state.constraints[to_addr] = state.u[from_addr]
+    else
+        set_submap!(state.constraints, to_addr, get_submap(state.u, from_addr))
+    end
+    nothing
+end
+
+function copy_proposal_to_model(state::JacobianPassState, from_addr, to_addr)
+    nothing
+end
+
+# call another involution function (NOTE: only the top-level function is actually technically an involution)
+
+macro call(ex)
+    MacroTools.@capture(ex, f_(args__)) || error("expected syntax: f(..)")
+    quote $(esc(f)).fn!($(esc(inv_state)), $(map(esc, args)...)) end
+end
+
+# apply 
+
+function apply_involution(involution::InvolutionDSLProgram, trace, u, proposal_args, proposal_retval)
+
+    # take the first pass through the involution, mutating first_pass_state
+    first_pass_state = FirstPassState(trace, u)
+    involution.fn!(first_pass_state, get_args(trace), proposal_args, proposal_retval)
+
```

Comment:
@marcoct If we call `update` here, rather than after `apply_involution` returns, can't we use the `discard` to automatically determine which continuous reads were `retained`, and construct `f_array` accordingly? That way, we wouldn't need to expose any special `retained` functionality to the user. Or maybe I'm missing something?

Reply:
@alex-lew That's interesting. I will look more closely into that.

Reply:
@alex-lew Yeah, you're right. We can just run `update` first. I made this change. Nice find!